Fix addressing issues in fabtests

sayantansur opened the pull request:
fabtests were assuming various things about addressing

1) Remove usage of getaddrinfo() from the tests themselves
- tests should provide client and server node names, and
fi_getinfo should do the right thing for address resolution

2) Rename 'src_addr' and 'dst_addr' in command line options
to 'src_node' and 'dst_node'. Removed FI_SOCKADDR_IN requirement.
- portable OFI programs shouldn't assume any format of address.
interfaces can be chosen with names.

e.g. to use IPoIB, use node-ib0, vs. passing in IP address of ib0
interface.

3) fix bug in ud_pingpong.c where hints->dest_addr was used for
av_insert, vs fi->dest_addr
- fi->dest_addr was obtained from getinfo(), vs. hints which is
a hardcoded address (also portable OFI programs should use
fi_getinfo for getting correct address structures)

Signed-off-by: Sayantan Sur <sayantan.sur@intel.com>

Tested with sockets provider, seems to be working.

@jithinjosepkl @shantonu @goodell @jsquyres @shefty @hppritcha Please review!

jsquyres commented:
@sayantansur Can you explain why you renamed src_addr -> src_node (etc.)?  "Addr" is better because it implies a specific interface, vs. "node" which users tend to conflate with a DNS/resolvable hostname.

sayantansur commented:
The data structures are developer-facing, not user facing. As long as the command line arguments for help say the right thing, we can avoid confusing users.

Secondly, address and node are different things in OFI. src_addr, dst_addr are defined as void * in fi_info, interpreted by an enum. 'node' is a string that is passed as a parameter to fi_getinfo. Tests were passing in src_addr, dst_addr as a fi_getinfo() parameter that is called 'node' :)

Therefore to unify things, we should use addr where we imply an OFI address (which can be a provider specific structure - say sockaddr_ib), and node where we imply a string that can be looked up by a provider and construct a fabric appropriate structure.

(FWIW, users will like using DNS resolvable names for interfaces vs IP addresses in command line args. Certainly the fabtests should be capable of taking in node names and not require IP addresses).

shefty commented:
How do these changes support the user entering both the source and destination addresses and ports?

sayantansur commented:
That functionality should be intact. Users can still supply src_node, and src_port. The port part was unchanged, and src_addr was just renamed to src_node.

shantonu commented on ported/libibverbs/rc_pingpong.c:
```diff
@@ -508,11 +508,15 @@ int main(int argc, char *argv[])
+	} else {
+		node = opts.dst_node;
+		service = opts.dst_port;
+	}
```
Not a big issue but can we put the common code for setting node, service, and flags in ft_read_addr_opts()? Every example seems to have the same code.

sayantansur replied:
yea, good point. i'll update.

sayantansur replied:
updated in 33e8a47.

shefty commented:
> That functionality should be intact. Users can still supply src_node, and
> src_port. The port part was unchanged, and src_addr was just renamed to
> src_node.

My question was how can the user provider BOTH the source and destination addresses and ports?  I.e. source node + source service + destination node + destination service.

sayantansur commented:
Talked with Sean. Will fix the PR soon.

zakattacktwitter commented:
Would be great to test this out on something other than the sockets provider and a single host. The ultimate test is to check this on *ALL* providers in a multi host environment (i.e. nodes are all physically different). Maybe I am missing something but it has been quite a struggle to debug how to use the same initialization code across different providers. Fabtests should be the canonical example. Thanks!

pmmccorm commented:
@zakattacktwitter Yes. That is the biggest TODO in the fabtests: make the initialization and shutdown code common across most or all tests. There might be an issue open for this.

sayantansur commented:
Sorry, I've been busy and haven't had the time to redo this PR. Will pick this up soon.

I can test on sockets and verbs providers.

goodell commented:
@sayantansur ping the PR once things are in a state you would like to see reviews and testing again.  Ben and I can take a look and try your changes out with the usnic provider.

sayantansur commented:
I updated the PR to incorporate Sean's comments. We need to provide the functionality that we start the client up and ask it to use a specific source address. The tests were using getaddrinfo to get the specific src address, and I just modified them to use fi_getinfo() instead, so it is more portable on OFI providers.

I haven't yet modified unit/av_test.c - that test still uses getaddrinfo() and sockaddr_in directly. I wasn't able to immediately understand the intent/functioning of the test - need some help in understanding it. Maybe there is a better way to write that test that it becomes portable.

@a-ilango can you help me verify verbs? I will ask others to verify after your comments.

a-ilango commented:
@sayantansur ok, I'll run these over verbs and report back.

sayantansur commented:
@a-ilango and I just tested. Seems to work as advertised.

@shantonu @shefty @goodell @j-xiong, can you verify?

shefty commented on common/shared.c:
```diff
@@ -80,47 +80,62 @@ static int getaddr(char *node, char *service, void **addr,
 		if (ret)
 			return ret;
-		*node = opts->dst_addr;
+		*node = opts->dst_node;
```
This sort of renaming ends up producing a huge amount of noise in a patch and needs to be in a separate commit.

shefty commented on pingpong/msg_pingpong.c:
```diff
@@ -442,11 +442,7 @@ static int client_connect(void)
 
-	ret = ft_getsrcaddr(opts.src_addr, opts.src_port, hints);
-	if (ret)
-		return ret;
```
What happened to this functionality?

shefty commented on common/shared.c:
```diff
@@ -80,47 +80,62 @@ static int getaddr(char *node, char *service, void **addr,
+		} else {
+			FT_ERR("addr allocation failed\n");
+			ret = EAI_MEMORY;
+		}
```
Is the intent that the addr and len be copied back into hints?  If so, I think you need to copy back the addr_format that the provider returns.  On a subsequent call to fi_getinfo(), a different provider than the one that returned the source address may interpret the address incorrectly.

shefty commented on pingpong/rdm_cntr_pingpong.c:
```diff
@@ -327,16 +327,10 @@ static int init_fabric(void)
+	ret = ft_read_addr_opts(&node, &service, hints, &flags, &opts);
+	if (ret) {
+		FT_PRINTERR("ft_read_addr_opts", ret);
+		return ret;
```
Please keep to 1 concept per patch.  This change looks unrelated to selecting the address.

shefty commented on pingpong/ud_pingpong.c:
```diff
@@ -405,11 +399,7 @@ static int client_connect(void)
-		goto err1;
-
-	ret = fi_av_insert(av, hints->dest_addr, 1, &remote_fi_addr, 0, NULL);
+	ret = fi_av_insert(av, fi->dest_addr, 1, &remote_fi_addr, 0, NULL);
```
This looks like an actual bug fix and should probably go in separately.

shefty commented on ported/libibverbs/rc_pingpong.c:
```diff
@@ -508,11 +508,12 @@ int main(int argc, char *argv[])
-		return 1;
+	if (rc) {
+		FT_PRINTERR("ft_read_addr_opts", rc);
+		return -rc;
```
Other places simply return the result of ft_read_addr_opts, but here we negate it.

shefty commented on common/shared.c:
```diff
@@ -80,47 +80,62 @@ static int getaddr(char *node, char *service, void **addr,
+		} else {
+			FT_ERR("addr allocation failed\n");
+			ret = EAI_MEMORY;
+		}
```
When would this function be called without FI_SOURCE set, and why?

shefty commented:
Conceptually, I think this moves in the right direction.  I would have liked to see a lot less noise in the patch, but I'm willing to accept the noise this time because of the number of files that need to be changed.  But please separate noise into its own patch.  I can't easily tell from reading the patch, but what I anticipate the code doing is this:

server applications: single call to fi_getinfo with source specified as node/service
client applications:
- if no source is specified, single call to fi_getinfo with dest specified as node/service
- if source given, call getaddr() to select source, second call to fi_getinfo with dest specified as node/service and source part of hints

Is this the resulting behavior from this patch?

shefty commented:
See issue #1117 for an alternate solution that may give us what we want in an easier way.  Note that that solution changes libfabric.  fabtests for compatibility reasons would likely need to support both options, so that it can run against old and new versions of libfabric.

sayantansur commented:
Yes, that was what the code was doing. The function ft_read_addr_opts decides whether the caller is server or client. If client && source provided -> call fi_getinfo() to get the source address and put it in the hints->src_addr, that is later used to call fi_getinfo() again with dest_addr set to server's address.

I'll separate out the patches.

shantonu commented:
Tested with Sockets provider and PSM. Seems to be working with these changes.

sayantansur commented:
Closing in favor of #290

goodell commented:
I'm slightly late to the party, but this PR did pass my usnic testing.  I'll check the replacement PRs as well.

sayantansur commented:
Thanks, Dave!
